plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergPageSourceProvider.java

```diff
@@ -319,14 +347,30 @@ private static ConnectorPageSource createOrcPageSource(
                         readType = column.getType();
                     }
 
+                    List<List<Integer>> fieldIdProjections = fileColumnsByIcebergId.isEmpty() ?
+                            projectionsByName.get(column.getBaseColumnIdentity().getName()) :
+                            projectionsByFieldId.get(column.getId());
+                    ProjectedLayout projectedLayout = IcebergProjectedLayout.createProjectedLayout(orcColumn, fieldIdProjections);
+
                     int sourceIndex = fileReadColumns.size();
                     columnAdaptations.add(ColumnAdaptation.sourceColumn(sourceIndex));
                     fileReadColumns.add(orcColumn);
                     fileReadTypes.add(readType);
+                    projectedLayouts.add(projectedLayout);
 
-                    Domain domain = effectivePredicateDomains.get(column);
-                    if (domain != null) {
-                        predicateBuilder.addColumn(orcColumn.getColumnId(), domain);
+                    for (Map.Entry<IcebergColumnHandle, Domain> domainEntry : effectivePredicateDomains.entrySet()) {
```

Comment:
Does the same issue apply to Hive connector?

Reply:
In Hive, for each read column we're first filtering out the domains from the `effectivePredicate` that matter to that column: https://github.com/trinodb/trino/blob/master/plugin/trino-hive/src/main/java/io/trino/plugin/hive/orc/OrcPageSourceFactory.java#L345

It's about the same, a linear filter over the predicate column for each base column.

@findepi we could simplify the code here a little now that we're only ever projecting base columns. I wrote `covers` to work even if both of the columns were projected subfields (a.b covers a.b.c), but now we know that we're only reading base columns, we could just compare `IcebergColumnHandle#getBaseColumnIdentity`

Reply:
>  we could just compare IcebergColumnHandle#getBaseColumnIdentity

that would allow us for a faster search

however, is there anything guaranteeing predicated columns are only base columns?
if we push `struct.field` projection down into Iceberg connector,
then we can subsequently push `struct.field > 0` filter into it as well, right?

Reply:
The predicate columns are not guaranteed to be base columns, but the read column is. So we'd be comparing `readColumn.getColumnIdentity().equals(predicateColumn.getBaseColumnIdentity())` to filter out predicates that belong to that read column.

I'm adding some `verify`s to assert that the read column is always a base column. It's not super clear that's the case right now, but it is.